src/crimson/osd/pg.cc

```diff
@@ -661,7 +661,24 @@ seastar::future<Ref<MOSDOpReply>> PG::handle_failed_op(
   }, load_obc_ertr::assert_all{ "can't live with object state messed up" });
 }
 
-seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
+crimson::errorator<crimson::ct_error::eagain>::future<> PG::rep_repair_primary_object(
+  Ref<MOSDOp> m,
+  const hobject_t& oid,
+  eversion_t& v) 
+{
+  ceph_assert(is_primary());
+  logger().debug("{}: {} peers osd.{}", __func__, oid, get_acting_recovery_backfill());
+  // Add object to PG's missing set if it isn't there already
+  if (!get_local_missing().is_missing(oid)) {
+    peering_state.force_object_missing(pg_whoami, oid, v);
+  }
+  // Send Pull&Push messages to OSD holding replicas and add object to the recovery map
+  auto [op, fut] = get_shard_services().start_operation<crimson::osd::UrgentRecovery>(
+		oid, v, this, get_shard_services(), m->get_min_epoch());
+  return crimson::ct_error::eagain::make();
```

Comment:
no need to return an error here.

Reply:
I was trying to achieve a similar behavior to what we currently have in ceph-osd, take a look here: https://github.com/ceph/ceph/blob/0aba110696b956581f7777147da27db589343f33/src/osd/PrimaryLogPG.cc#L15056

Reply:
@athanatos what do you think?

Reply:
What is it that ceph-osd does that you are trying to emulate?

Reply:
Since we need to return `eagain` from `do_osd_ops()` in this case, I thought it made sense to have the `eagain` return statement inside `rep_repair_primary_object()` and then do something like `return rep_repair_primary_object()` from `do_osd_ops()`.

Reply:
I don't much like it.  This is the success outcome for this function -- the return values should make sense in isolation without needing to examine all users.  I think do_osd_ops should translate the success into an EAGAIN return because it's specifically do_osd_ops that had an abnormal outcome.